### ISSUE TITLE
Handle MIDI timing in case of lag

### ISSUE THREAD

**Issue by 1j01** [1j01/midi-recorder#8]
Currently, start/end times are determined from a timer separate from the MIDI clock; it uses `performance.now()` so it records the time the event was received, or rather *handled*. This is the only straightforward way to do it, because I don't think the Web MIDI API gives you timestamps!

But what happens is if the page lags, the notes build up and all appear to start/end once the lag is over. (and are recorded as such.)
This is not good!

Investigate:
- Is it possible to access WebMIDI from a WebWorker?
- Is there any better timer to use, like the clock signals in MIDI? Can we access that and use it to correct for lag?

**Comment by 1j01** [comment]
In https://github.com/1j01/midi-recorder/commit/71ad9073c9c94c3000997768220afe08d2422e9d, I made it use timestamps from the MIDI events, when available, which may fix this.

**Comment by 1j01** [comment]
I think it's okay now.

**Comment by aDu** [comment]
I rather it use timestamps, I recorded using this and sadly later on I found the MIDI did not align with my performance.

Especially when I play faster pieces with several events per second and have other programs open, please can this issue be re-opened?

I do not have this issue with other MIDI recorders. The timings drifted by about 1s every 30 seconds, which was significant.

**Comment by 1j01** [comment]
I can reopen the issue, but I'm not sure what we can do about it, short of recommending people not use the app - and perhaps creating a native version - which I would be willing to do if it's the only way to capture performances accurately.
Like I said in a comment above, I did make it use timestamps from the MIDI events if available.

When you say the timings drifted, do you mean a BPM issue (time scaling)? Or are you noticing relative timing issues? Is it like gaps in the recording? (I don't know if you can tell that.)

We can test this scientifically by playing a steady rhythm, ideally generated by a MIDI device (perhaps a demo song built into a MIDI keyboard, with lots of fast notes), and artificially creating system lag.

I've recorded thousands of MIDIs with this app, and only listened back to a few of them, so I'm rather invested in this myself - but also afraid to find out. If the app truly has precision problems, it'd be better to rip off the bandaid and find out sooner rather than later.

If you'd be willing to perform such an investigation, I'd be grateful.

**Comment by aDu** [comment]
> Like I said in a comment above, I did make it use timestamps from the MIDI events if available.

Interesting, I have no idea why then. The drifting happened at certain points suddenly, I used Roli MIDI Piano. Maybe it's a bug with the piano, I will need to test it more. I wouldn't have noticed normally but I typically use the MIDI to edit visuals for my posts e.g. https://www.instagram.com/p/DUMvi7bCJsl/ so when there's misalignment I notice while editing.

In the middle of the piece there was sudden drifting.

I sliced part of the MIDI I used and fixed for one of my videos (I fixed it using Transkun Wav to Midi as luckily I had the Wav recording and then used a script to fix the onset time of the recorded MIDI)

(Uploading as txt as Github won't let me upload .mid)

[MIDI Recording (drifted time, sliced).midi.txt](https://github.com/user-attachments/files/25157511/MIDI.Recording.drifted.time.midi.txt)
[Manually fixed recording, sliced.mid.txt](https://github.com/user-attachments/files/25157512/Manually.fixed.recording.mid.txt)
[2026-01-31T180513Z att 5.midi (original unsliced file I downloaded, i used the last part for the video).txt](https://github.com/user-attachments/files/25157588/2026-01-31T180513Z.att.5.midi.txt)

I have to say the Roli MIDI does have a lot of bugs (including bugs that crash restarts Windows that affect multiple users) so maybe it's not related to your program.

If no-one else is reproducing this bug then it can probably be closed. All I did was play using the Roli Airwave (I did have multiple MIDI devices connected to the website as input, maybe it interfered but no idea, and the Airwave that tracks hand movement in the air counted as Midi Input).
